Custom validation message configuration in form definition

gigxz opened the pull request.

gigxz committed:
Custom validation message configuration in form definition

gigxz committed:
Update schema

eanders commented:
@gigxz are we waiting to merge this? https://github.com/greenriver/hmis-warehouse/pull/2854

gigxz committed:
Merge branch 'release-68' into gig/income-validation

gigxz committed:
Checkout income form def

gigxz commented:
It can be merged, but I removed the income fragment change (https://github.com/greenriver/hmis-warehouse/pull/2854/commits/c4297828b93df91c2f14bb9ceec981e3e29210d1) so it won't be used yet.

The HUD rule in question is "if YES is selected for insurance, at least one insurance must be picked." (ditto for income and benefits).

I changed my mind about where that validation rule should be defined and evaluated. Given that it is a HUD requirement, it's something we should validate on the model before saving it. 

We _could_ support ensuring all records match this rule in a migration because we could select "other => unspecified". Even if we don't want to put it in AR `validate`, we could put it in `hmis_validate` (the "custom" validator function) to ensure that we are validating _on the record_ rather than on the JSON form values.

HUD-compliance validation rules should be pushed as far down as possible. It's OK to duplicate them in the FormDefinition (if we really want a visual cue on blur) but it shouldn't be necessary. That is the spirit behind changes proposed in https://github.com/greenriver/hmis-warehouse/pull/2888 (haven't requested review yet, but I just got all tests passing. Just need to test frontend e2e and it should be ready).
